modules/core/src/main/scala/iota/internal/toolbelts.scala

```diff
@@ -163,4 +223,30 @@ private[internal] sealed abstract class IotaCommonToolbelt {
       case t => Left(s"unexpected type $t ${showRaw(t)} when destructuring CopK $tpe")
     }
 
+  class TypeListBuilder private[IotaCommonToolbelt](
+    consTpe: Type,
+    nilTpe: Type
+  ) {
+    final def apply(tpes: List[Type]): Type =
+      tpes.reverse.foldLeft(nilTpe)((t, h) => makeCons(h, t))
```

Comment:
This can be `tpes.foldRight(nilTpe)(makeCons)`.

I think `List#foldLeft` is implemented as `reverse` and `foldLeft`, so it should do exactly the same.